ffi/derive/src/convert.rs

```diff
@@ -20,140 +27,182 @@ pub fn derive_ffi_type(mut input: DeriveInput) -> TokenStream {
         Data::Enum(item) => {
             let repr = find_attr(&input.attrs, "repr");
 
-            if item.variants.is_empty() {
-                abort!(name, "uninhabited enum's cannot be instantiated");
-            }
-
             if is_fieldless_enum(&item) {
+                if item.variants.len() == 1 {
+                    // NOTE: one-variant fieldless enums have representation of ()
+                    return derive_ffi_type_for_opaque_item(name, &input.generics);
+                }
+                if without_repr(&input.attrs) {
+                    abort!(name, "Fieldless enum missing #[repr(int)]");
+                }
                 if item.variants.iter().any(|v| v.discriminant.is_some()) {
                     abort!(
                         name,
-                        "fieldless enums with explicit discriminants are prohibited"
+                        "Fieldless enums with explicit discriminants are prohibited"
                     );
                 }
+
                 derive_ffi_type_for_fieldless_enum(&input.ident, &item, &repr)
             } else {
                 let local = !is_non_local(&input.attrs);
                 derive_ffi_type_for_data_carrying_enum(&input.ident, input.generics, &item, local)
             }
         }
-        Data::Struct(_) => {
-            if is_owning(&input.attrs, &input.data) {
-                abort!(
-                    name,
-                    "Structure contains raw pointers. If the pointers don't own the data, attach `#[ffi_type(unsafe {non_owning})`"
-                )
-            }
+        Data::Struct(ref item) => {
+            let ffi_type_impl = derive_ffi_type_for_repr_c(&input);
+
+            let repr_c_impl = {
+                let predicates = &mut input.generics.make_where_clause().predicates;
```

Comment:
Weird function to me, but that's on `syn` side, I understand it.
Because `make_` sounds like it returns an object by value. And by reading it looks like you take a mutable reference to a temporary value, which is forbidden in Rust

Reply:
It's even more weird that `syn` allows you to insert new clauses like that

Reply:
it's `syn` API, I've used it a lot of times